File: C11/백준/Gold/9466. 텀 프로젝트/텀 프로젝트.c
#include <stdio.h>
#include <string.h>

#define MAX 100001

int student[MAX];
int visited[MAX];
int finished[MAX];
int count;

void dfs(int x) {
    visited[x] = 1;
    int next = student[x];

    if (!visited[next]) {
        dfs(next);
    } else if (!finished[next]) {
        count++;
        for (int i = next; i != x; i = student[i]) {
            count++;
        }
    }

    finished[x] = 1;
}

int main() {
    int t, n;
    scanf("%d", &t);

    while (t--) {
        scanf("%d", &n);

        for (int i = 1; i <= n; i++) {
            scanf("%d", &student[i]);
        }

        memset(visited, 0, sizeof(int) * (n + 1));
        memset(finished, 0, sizeof(int) * (n + 1));
        count = 0;

        for (int i = 1; i <= n; i++) {
            if (!visited[i]) {
                dfs(i);
            }
        }

        printf("%d\n", n - count);
    }

    return 0;
}

File: Java/백준/Gold/9466. 텀 프로젝트/텀 프로젝트.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {
    static int[] student;
    static boolean[] visited;
    static boolean[] finished;
    static int count;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            int n = Integer.parseInt(br.readLine());
            student = new int[n + 1];
            visited = new boolean[n + 1];
            finished = new boolean[n + 1];
            count = 0;

            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 1; i <= n; i++) {
                student[i] = Integer.parseInt(st.nextToken());
            }

            for (int i = 1; i <= n; i++) {
                if (!visited[i]) {
                    dfs(i);
                }
            }

            System.out.println(n - count);
        }
    }

    static void dfs(int x) {
        visited[x] = true;
        int next = student[x];

        if (!visited[next]) {
            dfs(next);
        } else if (!finished[next]) {
            count++;
            for (int i = next; i != x; i = student[i]) {
                count++;
            }
        }

        finished[x] = true;
    }
}